### PR TITLE
WIP, Switching MPS to use `SharedMemory`, adapt MPS<->CPU to COW. 

### PR DESCRIPTION
**Pull request by erpang007chenfs**
Fixes #ISSUE_NUMBER

Currently simple tests don't pass 
```
import torch
torch.set_future_lazy_clone(True)

cpu = torch.zeros((1, 2, 3), device=torch.device('cpu'))
mps = cpu.to("mps")

print(mps)
```
will result 
```
Traceback (most recent call last):
  File "/Users/chenfan_sun/workspace/pytorch/test_cow.py", line 67, in <module>
    main()
  File "/Users/chenfan_sun/workspace/pytorch/test_cow.py", line 63, in main
    verify_new_behavior()
  File "/Users/chenfan_sun/workspace/pytorch/test_cow.py", line 26, in verify_new_behavior
    print(mps)
  File "/Users/chenfan_sun/workspace/pytorch/torch/_tensor.py", line 464, in __repr__
    return torch._tensor_str._str(self, tensor_contents=tensor_contents)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/chenfan_sun/workspace/pytorch/torch/_tensor_str.py", line 705, in _str
    return _str_intern(self, tensor_contents=tensor_contents)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/chenfan_sun/workspace/pytorch/torch/_tensor_str.py", line 625, in _str_intern
    tensor_str = _tensor_str(self, indent)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/chenfan_sun/workspace/pytorch/torch/_tensor_str.py", line 357, in _tensor_str
    formatter = _Formatter(get_summarized_data(self) if summarize else self)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/chenfan_sun/workspace/pytorch/torch/_tensor_str.py", line 137, in __init__
    tensor_view = tensor.reshape(-1)
                  ^^^^^^^^^^^^^^^^^^
RuntimeError: allocator_ INTERNAL ASSERT FAILED at "/Users/chenfan_sun/workspace/pytorch/c10/core/StorageImpl.h":76, please report a bug to PyTorch. For resizable storage, allocator must be provided
```

### CODE REVIEW

**Comment by kurtamohler** [comment]
I notice that this PR is based off of my in-progress PR #126129. Just my two cents, I think it would be better to base off of main. I don't think your changes really require anything that my PR adds. You may get this PR working before my PR is finished, and it would probably be ideal for you to be able to merge it at that point rather than having to wait for my PR to go through first. Depending on whose PR is ready first, I think fixing merge conflicts will be relatively straightforward

**Comment by erpang007chenfs** [comment]
Sounds good. I will cherry pick and change the use main as my base branch.

**Comment by erpang007chenfs** [comment]
I do believe that I will need functions such as `_lazy_clone_future` in order to make the MPS<->CPU as COW to work, unless you think otherwise.

File: aten/src/ATen/native/AutogradComposite.cpp [hunk 1]
```diff
@@ -117,6 +123,50 @@ static Tensor _lazy_clone_impl(Tensor const& self, bool future) {
...
+    if (allocator == nullptr) {
+      std::cout << "Allocator is nullptr" << std::endl;
+    } else if (allocator == at::mps::GetMPSAllocator()) {
+      std::cout << "Allocator is MPS allocator" << std::endl;
+    } else {
+      std::cout << "Allocator is CPU allocator" << std::endl;
+    }
+
+    storage->set_allocator(c10::GetAllocator(device.value().type()));
+    storage->_mutable_data_ptr_no_checks().unsafe_set_device(device.value());
```
Comment (kurtamohler) [inline aten/src/ATen/native/AutogradComposite.cpp:164]: Maybe at first, try just setting the allocator and not the device. Does that at least fix the error you're getting?

**Comment by kurtamohler** [comment]
> I do believe that I will need functions such as `_lazy_clone_future` in order to make the MPS<->CPU as COW to work, unless you think otherwise.

In the main branch, `_lazy_clone` is the same as `_lazy_clone_future` from my PR.

Although I guess I should ask, do we need to do a deprecation cycle to change the behavior of `cpu_tensor.to('mps')`? If so, then your changes really would depend on my PR

File: aten/src/ATen/native/AutogradComposite.cpp (hunk 1 above)
Comment (kurtamohler) [reply aten/src/ATen/native/AutogradComposite.cpp:164]: I would expect this to fail as well:

```python
import torch
torch.set_future_lazy_clone(True)

cpu = torch.zeros((1, 2, 3), device=torch.device('cpu'))
mps0 = cpu.to("mps")
mps1 = mps0._lazy_clone()
```

**Pull request closed without merging**